### PR TITLE
[Model] Add classification Task with Qwen2ForSequenceClassification 

### LINKED ISSUE #8700: [Feature]: Support for Seq classification/Reward models
### 🚀 The feature, motivation and pitch

Verifier/reward models are going to be very important moving forward for building:
- High quality synthetic data pipelines
- Verifying model reasoning
- Multi agent systems

Could we add support for sequence classification models like Skywork/Skywork-Reward-Llama-3.1-8B

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by kakao-kevin-us**
FILL IN THE PR DESCRIPTION HERE

FIX #8700 (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

Key changes:
- Implement Qwen2ForSequenceClassification model architecture
- Add model registration for sequence classification variant
- Include integration tests for the new model variant
- Update documentation to reflect new sequence classification capability
- Usecase is similar like #8896
```bash
python -m vllm.entrypoints.openai.api_server \
	--model jason9693/Qwen2.5-1.5B-apeach \
	--trust-remote-code \
	--served-model-name Qwen2.5-1.5B-APEACH \
	--port 8080 \
	--tensor-parallel-size 8 \
	--enforce-eager
```

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -271,6 +271,16 @@ def __init__(
...
+            config = AutoConfig.from_pretrained(
+                model_name,
+                torch_dtype=torch_dtype,
+                trust_remote_code=True,
+            )
+            arch = config.architectures 
+            if len(arch) > 0:
+                cls_type = arch[0].split("For")[-1]
+                auto_cls = eval(f"AutoModelFor{cls_type}")
+
```
Comment (DarkLight1337) [inline tests/conftest.py:283]: This looks unsafe. We can already specify the class to use using `auto_cls` argument, so there is no need to do this.

File: vllm/model_executor/models/qwen2_cls.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                                   attn_metadata, intermediate_tensors)
+        logits, _ = self.score(hidden_states)
+        return logits
+
+    def pooler(
+        self,
+        hidden_states: torch.Tensor,
+        pooling_metadata: PoolingMetadata,
+    ) -> Optional[PoolerOutput]:
+        return self._pooler(hidden_states, pooling_metadata)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen2_cls.py:113]: I think it makes more sense to place the softmax here instead of inside pooler.

File: tests/conftest.py [hunk 3]
```diff
@@ -334,6 +344,18 @@ def get_inputs(
 
         return all_inputs
 
+    def classify(self, prompts: List[str]) -> List[str]:
+        # output is final logits
+        all_inputs = self.get_inputs(prompts)
+        outputs = []
+        print(f"model: {self.model}")
```
Comment (DarkLight1337) [inline tests/conftest.py:351]: There are still a bunch of print statements in the code. Please remove them before finalizing this PR.

**Review by DarkLight1337 — COMMENTED** [review]
Thanks for adding this! Some initial comments.

File: tests/conftest.py (hunk 1 above)
Comment (jason9693) [reply tests/conftest.py:283]: Thank you for your comments.
I want to take your opinion related to this issue.

the `auto_cls` is registered as "AutoModelForCausalLM" by default.
However, in this scene, we want to register as "AutoModelForSequenceClassification".
What do you think about the best option for this situation & context?

File: vllm/model_executor/models/qwen2_cls.py (hunk 2 above)
Comment (jason9693) [reply vllm/model_executor/models/qwen2_cls.py:113]: Thank you for your opinion. 
I agree with your way.

I'll commit this with followed by yours.

File: tests/conftest.py (hunk 3 above)
Comment (jason9693) [reply tests/conftest.py:351]: I just pushed it.

File: tests/conftest.py (hunk 1 above)
Comment (DarkLight1337) [reply tests/conftest.py:283]: You can manually pass `auto_cls=AutoModelForSequenceClassification` when calling `hf_runner`.

Comment (DarkLight1337) [reply tests/conftest.py:283]: For example, the BART tests use the code

```
    with hf_runner(model, dtype=dtype,
                   auto_cls=AutoModelForSeq2SeqLM) as hf_model:
```

In your tests, you can use `AutoModelForSequenceClassification` instead.

Comment (jason9693) [reply tests/conftest.py]: Thank you for your suggestion.
This logic has been commited.

**Comment by jason9693** [comment]
@DarkLight1337  All of issues you commented was resolved.

File: .github/workflows/actionlint.yml [hunk 4]
```diff
@@ -28,7 +28,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: "Checkout"
-        uses: actions/checkout@eef61447b9ff4aafe5dcd4e0bbf5d482be7e7871 # v4.2.1
+        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
```
Comment (DarkLight1337) [inline .github/workflows/actionlint.yml:31]: Please remove these changes which are unrelated to this PR.

Comment (kakao-kevin-us) [reply .github/workflows/actionlint.yml]: I just removed

File: tests/models/decoder_only/language/test_cls_models.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+"""Compare the outputs of HF and vLLM when using greedy sampling.
+
+This test only tests small models. Big models such as 7B should be tested from
+test_big_models.py because it could use a larger instance to run tests.
+
+Run `pytest tests/models/test_models.py`.
+"""
```
Comment (DarkLight1337) [inline tests/models/decoder_only/language/test_cls_models.py:7]: Please either update this docstring to be accurate, or remove it.

File: vllm/model_executor/models/qwen2_cls.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                # Remapping the name of FP8 kv-scale.
+                name = maybe_remap_kv_scale_name(name, params_dict)
+                if name is None:
+                    continue
+                if is_pp_missing_parameter(name, self):
+                    continue
+                param = params_dict[name]
+                weight_loader = getattr(param, "weight_loader",
+                                        default_weight_loader)
+                weight_loader(param, loaded_weight)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen2_cls.py:157]: Can you try using `AutoWeightsLoader` to remove the need for this boilerplate code?

**Comment by DarkLight1337** [comment]
To improve visibility, I also suggest adding a new entry to the Supported Models page of the docs.

File: vllm/model_executor/models/qwen2_cls.py (hunk 6 above)
Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py:157]: I can try this in this time.
However, how about dealing with 
`Reward Model` simultaneously in other pull request?

I think it would be better to handle with other pull request for runtime-wise stability.

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_cls.py:157]: > However, how about dealing with Reward Model simultaneously in other pull request?

Which pull request are you referring to?

Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py:157]: > > However, how about dealing with Reward Model simultaneously in other pull request?
> 
> Which pull request are you referring to?

I'll open other request for this issue.

**Comment by kakao-kevin-us** [comment]
> To improve visibility, I also suggest adding a new entry to the Supported Models page of the docs.

I just added docs

File: vllm/model_executor/models/qwen2_cls.py (hunk 6 above)
Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py]: @DarkLight1337  Sorry, I think I'm confused.
I just committed for this.

File: tests/models/decoder_only/language/test_cls_models.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+"""Compare the outputs of HF and vLLM when using greedy sampling.
```
Comment (DarkLight1337) [inline tests/models/decoder_only/language/test_cls_models.py:1]: Can we move this to a new directory (similar to the one for embedding models), and update `test_pipeline.yaml` so that this is tested under Other Models Tests?

Comment (kakao-kevin-us) [reply tests/models/decoder_only/language/test_cls_models.py]: Just moved under embeddings

File: vllm/model_executor/models/qwen2_cls.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+class ReLU(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+        self.activation = nn.ReLU()
+
+    def forward(self, input):
+        input, _ = input
+        return self.activation(input)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen2_cls.py:34]: Looks like this is unused. Can we delete this?

File: vllm/model_executor/models/qwen2_cls.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
+# coding=utf-8
+# Adapted from
+# https://huggingface.co/Qwen/Qwen2.5-Math-RM-72B/blob/main/modeling_qwen2_rm.py
+# Copyright 2024 The Qwen team.
+# Copyright 2023 The vLLM team.
+"""Inference-only Qwen2-RM model compatible with HuggingFace weights."""
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen2_cls.py:6]: We should update this

File: vllm/model_executor/models/qwen2_cls.py (hunk 8 above)
Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py]: Ok, I see

File: vllm/model_executor/models/qwen2_cls.py (hunk 9 above)
Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py]: Just updated

**Comment by DarkLight1337** [comment]
It looks like the model test failed. PTAL.

File: vllm/model_executor/models/qwen2_cls.py (hunk 2 above)
Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py]: @DarkLight1337  this cause the runtime test failure.
Do you have any idea about this?

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_cls.py]: I think it's because `PoolerOutput` itself is not a tensor. Hmm... maybe we have to revert back to applying softmax inside the pooler.

Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py]: @DarkLight1337 I see. I'll do this now

Comment (kakao-kevin-us) [reply vllm/model_executor/models/qwen2_cls.py:102]: @DarkLight1337  Just pushed. Lets waiting the test results.

**Review by DarkLight1337 — APPROVED** [review]
The model test has passed, thanks for your time!

**Comment by DarkLight1337** [comment]
Please merge from main to fix the CI failures.

**Comment by kakao-kevin-us** [comment]
> Please merge from main to fix the CI failures.

I just rebased on main.
Let's waiting for the test

**Comment by corbt** [comment]
Very cool!

Is this compatible with the [LoRA adapters](https://docs.vllm.ai/en/v0.5.5/models/lora.html) functionality in vLLM? I'd love to be able to deploy a base `Qwen2ForSequenceClassification` and then have many different reward-model variants. The variants would have different LoRA layers and then separate `score` layers as well.

**Comment by kakao-kevin-us** [comment]
> Very cool!
> 
> Is this compatible with the [LoRA adapters](https://docs.vllm.ai/en/v0.5.5/models/lora.html) functionality in vLLM? I'd love to be able to deploy a base `Qwen2ForSequenceClassification` and then have many different reward-model variants. The variants would have different LoRA layers and then separate `score` layers as well.

It was tested on the CI, but not sure as I'm not tested on my local server.
If it is broken, you can merge the lora weight and run on the vLLM
https://huggingface.co/docs/diffusers/main/using-diffusers/merge_loras

**Comment by tcapelle** [comment]
Is this supported in Docker?
I am getting:
```
ValueError: Model architectures ['Qwen2ForSequenceClassification'] are not supported for now. Supported architectures: ['AquilaModel', 'AquilaForCausalLM', 'ArcticForCausalLM', 'BaiChuanForCausalLM', 'BaichuanForCausalLM', 'BloomForCausalLM', 'CohereForCausalLM', 'DbrxForCausalLM', 'DeciLMForCausalLM', 'DeepseekForCausalLM', 'DeepseekV2ForCausalLM', 'ExaoneForCausalLM', 'FalconForCausalLM', 'GemmaForCausalLM', 'Gemma2ForCausalLM', 'GPT2LMHeadModel', 'GPTBigCodeForCausalLM', 'GPTJForCausalLM', 'GPTNeoXForCausalLM', 'GraniteForCausalLM', 'GraniteMoeForCausalLM', 'InternLMForCausalLM', 'InternLM2ForCausalLM', 'JAISLMHeadModel', 'JambaForCausalLM', 'LlamaForCausalLM', 'LLaMAForCausalLM', 'MambaForCausalLM', 'MistralForCausalLM', 'MixtralForCausalLM', 'QuantMixtralForCausalLM', 'MptForCausalLM', 'MPTForCausalLM', 'MiniCPMForCausalLM', 'MiniCPM3ForCausalLM', 'NemotronForCausalLM', 'OlmoForCausalLM', 'OlmoeForCausalLM', 'OPTForCausalLM', 'OrionForCausalLM', 'PersimmonForCausalLM', 'PhiForCausalLM', 'Phi3ForCausalLM', 'Phi3SmallForCausalLM', 'PhiMoEForCausalLM', 'Qwen2ForCausalLM', 'Qwen2MoeForCausalLM', 'RWForCausalLM', 'StableLMEpochForCausalLM', 'StableLmForCausalLM', 'Starcoder2ForCausalLM', 'SolarForCausalLM', 'XverseForCausalLM', 'BartModel', 'BartForConditionalGeneration', 'Gemma2Model', 'MistralModel', 'Qwen2ForRewardModel', 'Phi3VForCausalLM', 'Blip2ForConditionalGeneration', 'ChameleonForConditionalGeneration', 'ChatGLMModel', 'ChatGLMForConditionalGeneration', 'FuyuForCausalLM', 'InternVLChatModel', 'LlavaForConditionalGeneration', 'LlavaNextForConditionalGeneration', 'LlavaNextVideoForConditionalGeneration', 'LlavaOnevisionForConditionalGeneration', 'MiniCPMV', 'MolmoForCausalLM', 'NVLM_D', 'PaliGemmaForConditionalGeneration', 'PixtralForConditionalGeneration', 'QWenLMHeadModel', 'Qwen2VLForConditionalGeneration', 'UltravoxModel', 'MllamaForConditionalGeneration', 'EAGLEModel', 'MedusaModel', 'MLPSpeculatorPreTrainedModel']
```
I don't see any SequenceClassification model support...

**Comment by DarkLight1337** [comment]
Which version are you using?
